### PR TITLE
Endpoint customizer refresh

### PR DESCRIPTION
**Pull request by kashifkhan0771**
### Description:
Updated Endpoint Customizer and Detectors using it.

This change separates configuration for using found URLs in the data and using known cloud URLs, while centralizing the logic for detectors to use in the `EndpointSetter` struct. The driving force for this change is that we have some URLs known ahead of time and some URLs known only at detection time, and we would like to configure both.

The embeddable `EndpointSetter` struct has a simplified interface: A detector calls `s.Endpoints(foundURLs...)` to get a slice of the endpoints to perform verification against.

A `CloudProvider` interface is also introduced in this PR for a detector to signal there is a cloud endpoint associated with it.

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [x] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/usage/install/#local-installation))?

### CODE REVIEW

File: pkg/detectors/artifactory/artifactory.go [hunk 1]
```diff
@@ -61,20 +64,24 @@ func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (result
...
+				s1.Verified = isVerified
+				s1.SetVerificationError(verificationErr, resMatch)
+			}
 
-		if verify {
-			client := s.getClient()
-			isVerified, verificationErr := verifyArtifactory(ctx, client, resURLMatch, resMatch)
-			s1.Verified = isVerified
-			s1.SetVerificationError(verificationErr, resMatch)
+			results = append(results, s1)
```
Comment (rgmz) [inline pkg/detectors/artifactory/artifactory.go:82]: Shouldn't a verified token break the loop? A token can't be valid on more than one instance, AFAIK.

Comment (kashifkhan0771) [reply pkg/detectors/artifactory/artifactory.go:82]: The idea is to capture the result of all URLs we verify against even if the verification passes or not.

Comment (abmussani) [reply pkg/detectors/artifactory/artifactory.go:82]: @rgmz I believe these are access token. If yes, then Form jfrog artifcatory [documentation](https://jfrog.com/help/r/jfrog-platform-administration-documentation/access-tokens), 

>  
> Cross-instance authentication
> 
> Access tokens can be used for authentication, not only by the instance or cluster where they were created but also for other instances and clusters that are all part of the same circle of trust.

Comment (rgmz) [reply pkg/detectors/artifactory/artifactory.go:82]: Interesting. Just be careful: Artifactory cloud has aggressive rate limiting. 

More than a few failed auth in a minute and you won't be able to hit it anymore, in my experience.

Comment (kashifkhan0771) [reply pkg/detectors/artifactory/artifactory.go:82]: @rgmz If Artifactory has any rate limiting, I believe that should be on the same endpoint. Here the endpoint will be changing for each request unless user pass same endpoint multiple times.

Comment (rosecodym) [reply pkg/detectors/artifactory/artifactory.go:82]: @kashifkhan0771 if an endpoint is found multiple times in a chunk, won't we query it once per occurrence? That could cause rate-limiting, right? It's unlikely but theoretically possible. (I don't know whether it's worth worrying about it.)

File: pkg/detectors/github/v1/github_old.go [hunk 2]
```diff
@@ -19,9 +19,10 @@ type Scanner struct{ detectors.EndpointSetter }
 var _ detectors.Detector = (*Scanner)(nil)
 var _ detectors.Versioner = (*Scanner)(nil)
 var _ detectors.EndpointCustomizer = (*Scanner)(nil)
+var _ detectors.CloudProvider = (*Scanner)(nil)
 
-func (Scanner) Version() int            { return 1 }
-func (Scanner) DefaultEndpoint() string { return "https://api.github.com" }
+func (Scanner) Version() int          { return 1 }
+func (Scanner) CloudEndpoint() string { return "https://api.github.com" }
```
Comment (zricethezav) [inline pkg/detectors/github/v1/github_old.go:25]: **Question**: Is there a reason we're changing `DefaultEndpoint` to `CloudEndpoint`? I feel like `DefaultEndpoint` is much easier to wrap my head around.

**Review by rosecodym — APPROVED** [review]

File: pkg/detectors/endpoint_customizer.go [hunk 3]
```diff
@@ -10,27 +10,43 @@ import (
...
+
+func (e *EndpointSetter) UseFoundEndpoints(enabled bool) {
+	e.useFoundEndpoints = enabled
+}
+
+func (e *EndpointSetter) Endpoints(foundEndpoints ...string) []string {
+	endpoints := e.configuredEndpoints
+	if e.useCloudEndpoint {
+		endpoints = append(endpoints, e.cloudEndpoint)
+	}
```
Comment (mcastorina) [inline pkg/detectors/endpoint_customizer.go:47]: Just thought of something: we should probably change this to only append if `e.cloudEndpoint != ""`:

```go
if e.useCloudEndpoint && e.cloudEndpoint != "" {
    endpoints = append(endpoints, e.cloudEndpoint)
}
```

File: pkg/detectors/artifactory/artifactory.go (hunk 1 above)
Comment (kashifkhan0771) [reply pkg/detectors/artifactory/artifactory.go:82]: @rosecodym good point! In that case I think it's good to add a check to verify that the endpoints should be unique only.

File: pkg/detectors/github/v1/github_old.go (hunk 2 above)
Comment (kashifkhan0771) [reply pkg/detectors/github/v1/github_old.go:25]: As per @mcastorina `EndpointCustomizer` was created for detectors that have cloud and on-prem endpoints. So we renamed it accordingly.

File: pkg/detectors/artifactory/artifactory.go (hunk 1 above)
Comment (kashifkhan0771) [reply pkg/detectors/artifactory/artifactory.go:82]: @rosecodym As per our logic for match URLs:
```
resURLMatch := ""
	for _, URLmatch := range URLmatches {
		if len(URLmatch) != 2 {
			continue
		}

		resURLMatch = strings.TrimSpace(URLmatch[1])
	}
```
We only pick one URL from the matches. I think a duplicate endpoint can only occur if user provide same endpoint multiple times. 
Shall we add a unique check for that or we can ignore it as that should be handled from user end?

File: pkg/detectors/endpoint_customizer.go (hunk 3 above)
Comment (kashifkhan0771) [reply pkg/detectors/endpoint_customizer.go]: Done @mcastorina

File: pkg/detectors/github/v1/github_old.go (hunk 2 above)
Comment (mcastorina) [reply pkg/detectors/github/v1/github_old.go:25]: `DefaultEndpoint` felt ambiguous and was really only defining a default *static* endpoint. Would `DefaultCloudEndpoint()` be a better name?

Comment (zricethezav) [reply pkg/detectors/github/v1/github_old.go:25]: `CloudEndpoint` is fine, thanks for the explanation

File: pkg/detectors/artifactory/artifactory.go [hunk 4]
```diff
@@ -52,6 +54,7 @@ func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (result
 		if len(URLmatch) != 2 {
 			continue
 		}
+
 		resURLMatch = strings.TrimSpace(URLmatch[1])
```
Comment (zricethezav) [inline pkg/detectors/artifactory/artifactory.go:58]: On a separate note, this detector appears to ignore all other URL matches other than the last one matched. https://github.com/trufflesecurity/trufflehog/blob/b2311b4ad2169f5d6c24e3ad136ce698ecfab042/pkg/detectors/mite/mite.go#L52 is how we should be handling this; nested loop w/ matches and urlMatches. Please correct me if I'm off basis here @mcastorina

Comment (mcastorina) [reply pkg/detectors/artifactory/artifactory.go:58]: That's correct - we noticed that oddity when we were updating this detector but decided it was out of scope for this PR.

Comment (zricethezav) [reply pkg/detectors/artifactory/artifactory.go:58]: Gotcha, in that case I think this is okay to merge

**Pull request merged**